### ISSUE TITLE
ImportError: Could not load any of ['hdf5.dll', 'hdf5dll.dll', 'pytables_hdf5.dll'], please ensure that it is installed in the package folder. Windows 10

### ISSUE THREAD

**Issue by Zachweiss20** [PyTables/PyTables#854]
python3 -m tables.tests.test_all
Traceback (most recent call last):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2032.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 185, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.8_3.8.2032.0_x64__qbz5n2kfra8p0\lib\runpy.py", line 111, in _get_module_details
    __import__(pkg_name)
  File "C:\Users\name\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.8_qbz5n2kfra8p0\LocalCache\local-packages\Python38\site-packages\tables\__init__.py", line 82, in <module>  
    raise ImportError(
ImportError: Could not load any of ['hdf5.dll', 'hdf5dll.dll', 'pytables_hdf5.dll'], please ensure that it is installed in the package folder.
PS C:\Users\name\Documents\GitHub\name>

**Comment by avalentino** [comment]
@Zachweiss20 please provide more information:
* windows version?
* python version ? (seems to be 3.8)
* PyTables version?
* Are you using conda?
* Are you using wheels?
* how have you installed hdf5 and pytabels?

**Comment by Zachweiss20** [comment]
Hey @avalentino, to answer your questions:
1. Windows 10
2. Python 3.8
3. I have installed tables through pip install tables;  it's version 3.6.1
4. Not using conda
5. Not using wheels
6. I installed h5py 3.1.0, and again from what i've read pytables is ~ tables, I've installed tables ver 3.6.1

**Comment by avalentino** [comment]
Thanks @Zachweiss20.
Can you please also tell us if during the installation process it has been downloaded ``tables-3.6.1-2-cp38-cp38-win_amd64.whl`` or ``tables-3.6.1.tar.gz``?

**Comment by Zachweiss20** [comment]
tables-3.6.1-2-cp38-cp38-win_amd64.whl
Ah, I guess that's the problem: Since i'm not using wheels this version isn't working for me?

**Comment by avalentino** [comment]
OK, if pip downloads ``tables-3.6.1-2-cp38-cp38-win_amd64.whl`` then you are using wheels as far as I can understand.
The wheel for python 3.8 should work out of the box, it included a copy of the hdf5 dll.

Looking better at your error log I see
```
...\site-packages\tables_init.py
```
I'm not familiar with this file, it is not part of the PyTables source code.
Do you have an idea where if comes form? 

> On a side note: Do you know if h5py will work with what tables needs?

h5py is a python wrapper of the HDF5 C library in the same way PyTables is.
Installing h5py or not should be totally irrelevant for PyTables AFAIK.

**Comment by avalentino** [comment]
@Zachweiss20 is this resolved?

**Comment by guiweber** [comment]
I'm seeing a similar issue when installing with `pip install tables` for Python 3.9.1. The file downloaded by pip is  `tables-3.6.1.tar.gz`. I've tried installing h5py by I still get the same error message when trying to install tables.

>     LINK : fatal error LNK1181: cannot open input file 'hdf5.lib'
>    \* Using Python 3.9.1 (tags/v3.9.1:1e5d33e, Dec  7 2020, 17:08:21) [MSC v.1927 64 bit (AMD64)]
>    \* USE_PKGCONFIG: False
>   .. ERROR:: Could not find a local HDF5 installation.
>     You may need to explicitly state where your local HDF5 headers and
>   library can be found by setting the ``HDF5_DIR`` environment
> variable or by using the ``--hdf5`` command-line option.

**Comment by guiweber** [comment]
Ok after researching a bit more it seems that's because no wheels are available for Python 3.9.

Would there by a way to improve the messaging on pip when no wheels are available for the python version?

**Comment by avalentino** [comment]
maybe the pip option ``--only-binary``

**Comment by amotl** [comment]
Hi @guiweber,

we are currently working on the infrastructure to build current wheels (see #872). However, we are seeing the same linker error specifically for Python 3.9 on Windows, see also https://github.com/PyTables/PyTables/issues/823#issuecomment-770460668.

Now, we are looking for support from people having a Windows development environment (either on hardware or via VM) who could fiddle with the respective routines within `setup.py` in order to get this going.

With kind regards,
Andreas.

**Comment by avalentino** [comment]
Not sure but the following could be related:
* https://docs.python.org/3/whatsnew/3.8.html#bpo-36085-whatsnew
* https://docs.python.org/3/library/os.html#os.add_dll_directory
* 
@amotl can you please give a try to ``os.add_dll_directory``?
It should be enough to add it at the very beginning of ``__init__.py`` and point to the wheel directory containing the hdf5 lib.

**Comment by amotl** [comment]
Hi Antonio,

thanks for those pointers. I will be happy to do some trial-and-error testing on them. However, I will be busy throughout the day, so you might want to integrate #872 beforehand. Then we can build upon that and add Python 3.9 / Windows support afterwards.

Also note that the build apparently works fine on Python 3.8 / Windows already. But you never know, maybe those changes are only needed for Python 3.9 because Python 3.8 still had some compatibility going on?

With kind regards,
Andreas.

**Comment by avalentino** [comment]
> thanks for those pointers. I will be happy to do some trial-and-error testing on them. However, I will be busy throughout the day, so you might want to integrate #872 beforehand. Then we can build upon that and add Python 3.9 / Windows support afterwards.

sure, no problem

**Comment by amotl** [comment]
Hi again,

I just wanted to come back to this in order to add a short note. I believe there might be two problems observed by Windows users.

1. Runtime errors like the one reported by @Zachweiss20 at https://github.com/PyTables/PyTables/issues/854#issue-778468683 as well as @Sophia409, @exrich and @e-dism at #824.
2. Compile-/Link-time errors like reported by @guiweber at https://github.com/PyTables/PyTables/issues/854#issuecomment-765462488 and @amotl at https://github.com/PyTables/PyTables/issues/823#issuecomment-770460668.

The suggestion by @avalentino at https://github.com/PyTables/PyTables/issues/854#issuecomment-770781554 will probably only solve the runtime error. We will see...

In order to add safeguards after mitigating those issues, we might think about improving the CI environment (#872) by actually **installing** the wheel packages into an isolated environment (where **no** build-time dependencies have been installed) and check the outcome.

With kind regards,
Andreas.

**Comment by justus-hildebrand** [comment]
Hey everyone, I have just run into this runtime error as well. I'm on a fresh install of python, here's my versions:

Windows 10 Pro 64-bit
python 3.8.9
pip 21.0.1
numpy 1.19.2
numexpr 2.7.3
tables 3.6.1

In case something with the cached wheel is broken I used an uncached install, here's the log:

```
C:\Users\User>pip install tables --no-cache-dir
Defaulting to user installation because normal site-packages is not writeable
Collecting tables
  Downloading tables-3.6.1-2-cp38-cp38-win_amd64.whl (3.1 MB)
     |████████████████████████████████| 3.1 MB 6.4 MB/s
Requirement already satisfied: numpy>=1.9.3 in c:\users\User\appdata\roaming\python\python38\site-packages (from tables) (1.19.2)
Requirement already satisfied: numexpr>=2.6.2 in c:\users\User\appdata\roaming\python\python38\site-packages (from tables) (2.7.3)
Installing collected packages: tables
  WARNING: The scripts pt2to3.exe, ptdump.exe, ptrepack.exe and pttree.exe are installed in 'C:\Users\User\AppData\Roaming\Python\Python38\Scripts' which is not on PATH.
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Successfully installed tables-3.6.1

C:\Users\User>python
Python 3.8.9 (tags/v3.8.9:a743f81, Apr  6 2021, 14:02:34) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tables
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\User\AppData\Roaming\Python\Python38\site-packages\tables\__init__.py", line 82, in <module>
    raise ImportError(
ImportError: Could not load any of ['hdf5.dll', 'hdf5dll.dll', 'pytables_hdf5.dll'], please ensure that it is installed in the package folder.
```
I've tried a couple of version combinations already, like up-/downgrading numpy (no success), and also tried installing an older version of tables, which gave me a different error.

@amotl You mentioned the suggestion by @avalentino, are there any branches/PRs attempting that solution already? I'd be glad to help out with testing!

**Comment by avalentino** [comment]
@justus-hildebrand thanks for reporting.
The relevant PR is #872 (please also refer to the discussion in  #823).
Any help would be very very welcome.

**Comment by justus-hildebrand** [comment]
With a last desperate attempt I finally managed to get it to work! **The solution for me was to install PyTables not with pip in user space but through pip run as Administrator.**

I _believe_ that it all has to do with environment variables: When installing `PyTables` from a non-admin pip, it is installed somewhere in pythons AppData directory (for me it was `c:\users\User\appdata\roaming\python\python38\site-packages`), which by default was not placed anywhere in my system's `PATH` variable (maybe I made a mistake when installing python here)

When installing as Administrator pip installs PyTables to `c:\program files\python38\lib\site-packages`, which **is** in `PATH` and allows python to locate the DLL when importing PyTables.

I think the right long term solution for this problem would be to place the DLL somewhere where python's file lookup expects DLLs to be, even when installing as User. Unfortunately you can't **know** that all environment variables are set correctly, but I reckon there should be **some** place in the user's directory that gets checked by python when importing packages that require DLLs. Sadly my knowledge of pythons package deployment systems is practically non-existent, else I could help out more with this.

@Zachweiss20, @Sophia409, @exrich and @e-dism, maybe you could check if this resolves your issues as well.

**Comment by golestani-rez** [comment]
hi to everyone
i had problem of converting csv to hdf5 format in python.
i read all valuable comments about 
ImportError: Could not load any of ['hdf5.dll', 'hdf5dll.dll', 'pytables_hdf5.dll']
and 
Missing optional dependency 'tables'. In pandas to_hdf and also 
LINK : fatal error LNK1181: cannot open input file 'hdf5.lib'
    * Using Python 3.8.2 (default, Aug 25 2020, 15:54:26) [MSC v.1900 64 bit (AMD64)]
    * Found cython 0.29.23
    * USE_PKGCONFIG: False
    * Found HDF5 using system PATH ('C:\Python38\DLLs\hdf5.dll')
    .. ERROR:: Could not find a local HDF5 installation.
       You may need to explicitly state where your local HDF5 headers and
       library can be found by setting the ``HDF5_DIR`` environment
       variable or by using the ``--hdf5`` command-line option.
windows 10, python 3.8.2, 

Solved for me: uninstall all prerequisite of tables 3.6.1 and install again including:
cython, numpy, Numexpr,raw-zlib, zlib-state, indexed-bzip2, blosc, bloscpack,lzo-indexer, python3-lzo-indexer, h5py, tables.
the problem was that i had these in multiple path.
i'm newbie and i did it empirically over one week.

**Comment by avalentino** [comment]
Fixed in #923.
